Add basic stub for characteristic

alexjurkiewicz opened the pull request:
https://characteristic.readthedocs.io/en/stable/

gvanrossum commented:
Did you get permission from the authors/developers of the package? See https://www.python.org/dev/peps/pep-0484/#the-typeshed-repo

hynek commented:
Permission granted and please look into using [attrs](http://www.attrs.org) instead. ;)

gvanrossum commented:
OK, but what we have here is incomplete -- it only defines one method. @alexjurkiewicz Can you make it more complete? (Maybe run stubgen or just transcribe the major API elements by hand.

alexjurkiewicz commented:
Thanks @hynek -- I apologise for my lack of legacy rework :D

I've added the remainder of the public API. How do you want the `pyi` file formatted? Some of the function definitions are extremely long, I've split them over multiple lines, but other files seem to use a single line...

gvanrossum commented:
This looks fine, but could you reformat using 4-space indents? That aspect of PEP 8 still applies... :-)

alexjurkiewicz commented:
whoops!

ambv commented:
@hynek, the irony of the `characteristic` tagline in this context does not escape me! 😆 

@gvanrossum, doesn't the use of class decorators like `immutable` or `with_repr` erase types with the current type definition? It seems to me those class decorators should rather be annotated as:
```
T = TypeVar('T')
def with_repr(attrs: Sequence[Union[AnyStr,Attribute]]) -> Callable[[T], T]: ...
```

gvanrossum commented on third_party/2and3/characteristic/__init__.pyi:
```diff
@@ -0,0 +1,34 @@
+from typing import Sequence, Callable, Union, Any, Optional, AnyStr, TypeVar, Type
+
+def with_repr(attrs: Sequence[Union[AnyStr,Attribute]]) -> Callable[..., Any]: ...
```
There should be a space after the comma between `AnyStr` and `Attribute` (here and below).

@ambv -- I'm surprised flake8 didn't catch this?

gvanrossum commented:
> @gvanrossum, doesn't the use of class decorators like immutable or with_repr erase types with the current type definition? It seems to me those class decorators should rather be annotated as:
> ```python
> T = TypeVar('T')
> def with_repr(attrs: Sequence[Union[AnyStr,Attribute]]) -> Callable[[T], T]: ...
> ```

Sadly that doesn't work yet: https://github.com/python/mypy/issues/1551

ambv commented:
> There should be a space after the comma between AnyStr and Attribute (here and below).
> 
> @ambv -- I'm surprised flake8 didn't catch this?
> 
Haha, this pull request predates flake8 integration. I only enabled it 3 days ago. Check it out locally and rebase, you'll see it fail the 'flake8' run.

alexjurkiewicz commented:
I'll be able to work on this again in early Jan, apologies for the delay

On Sat., 24 Dec. 2016, 8:48 am Łukasz Langa, <notifications@github.com>
wrote:

>
> > There should be a space after the comma between AnyStr and Attribute
> (here and below).
> >
> > @ambv -- I'm surprised flake8 didn't catch this?
> >
> Haha, this pull request predates flake8 integration. I only enabled it 3
> days ago. Check it out locally and rebase, you'll see it fail the 'flake8'
> run.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/typeshed/pull/771#issuecomment-269060905>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXKdbGRiS7D8th1obTAhfkkoGG1Yhmeks5rLGvogaJpZM4LNlSH>
> .
>

gvanrossum commented:
> Check it out locally and rebase, you'll see it fail the 'flake8' run.

I had done exactly that, and it passed. Then I found that I didn't run `pip3 install -U -r requirements-tests-py3.txt`. With that it finds these bugs.

I'm tempted to just fix the whitespace using GitHub's edit feature and merge.

gvanrossum commented:
Thanks! If you want to change more, just submit a new PR.

alexjurkiewicz commented:
Thanks for your help and time!

On Sat., 24 Dec. 2016, 11:12 am Guido van Rossum, <notifications@github.com>
wrote:

> Thanks! If you want to change more, just submit a new PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/python/typeshed/pull/771#issuecomment-269066310>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAXKdYHS0c8eJ9GVlm3EAZwXy20kY0Efks5rLI2TgaJpZM4LNlSH>
> .
>
